### PR TITLE
Added an Avatar for a user

### PR DESCRIPTION
**Pull request by tend2infinity**
This will close issue #35 
I have created another model for Avatar properties and linked that with the profile model. While on the frontend side I've used the required states and updated the state as well. 
Reference Video:

https://user-images.githubusercontent.com/61948033/118573443-33677280-b7a0-11eb-8d88-90501f22546e.mp4

### CODE REVIEW

File: client/src/components/profile/profile.js [hunk 1]
```diff
@@ -136,7 +141,7 @@ class Profile extends Component {
                 <div>
                   <img
                     className="profile-page__user-img"
-                    src="http://ssl.gstatic.com/accounts/ui/avatar_2x.png"
+                    src={`https://avataaars.io/?avatarStyle=Circle&topType=${avatar.top}&accessoriesType=${avatar.accessories}&hairColor=${avatar.hairColor}&facialHairType=${avatar.facialColor}&clotheType=${avatar.clothes}&eyeType=${avatar.eyes}&eyebrowType=${avatar.eyebrow}&mouthType=${avatar.mouth}&skinColor=${avatar.skin}&clotheColor=${avatar.clothColor}`}
```
Comment (fave77) [inline client/src/components/profile/profile.js]: Use the `avataaars` npm package to render the profile pic, like so:

```bash
<Avatar
  avatarStyle=''
  topType=''
  accessoriesType=''
  facialHairType=''
  clotheType=''
  eyeType=''
  eyebrowType=''
  mouthType=''
  skinColor=''
/>
```

File: server/src/models/avatar.model.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require('mongoose')
```
Comment (fave77) [inline server/src/models/avatar.model.js]: Do not create a separate model for `Avatar`

File: server/src/controllers/user.controller.js [hunk 3]
```diff
@@ -28,14 +29,27 @@ const registerUser = async (fullname, username, password, email, status, lichess
     salt: salt,
     status: status
   });
+  const newAvatar = new Avatar({
```
Comment (fave77) [inline server/src/controllers/user.controller.js]: Please pass these values from the front-end only, and make the user-controller as loosely coupled as possible.

File: client/src/components/profile/editprofile.js [hunk 4]
```diff
@@ -187,6 +274,240 @@ class EditProfilePopup extends Component {
               {alertFailure.msg}
             </Alert>
           }
+          <img
+            className="edit-profile-popup__user-img"
+            src={`https://avataaars.io/?avatarStyle=Circle&topType=${top}&accessoriesType=${accessories}&hairColor=${hairColor}&facialHairType=${facialHair}&clotheType=${clothes}&eyeType=${eyes}&eyebrowType=${eyebrow}&mouthType=${mouth}&skinColor=${skin}&clotheColor=${clothColor}`}
+            alt="your current profile"
+            />
+
+          {/* Avatar Properties */}
```
Comment (fave77) [inline client/src/components/profile/editprofile.js]: Create a separate file for edit-avatar inside `profile` folder, and refactor all the avatar-related boilerplate. You can also make the state separate as well, by making a completely different button for the edit avatar `pencil icon` attached to the profile pic.

So, Edit Profile will only edit the full name, country, etc, whereas Pencil Icon (edit avatar button attached next to profile pic) will change only avatar.

Both will hit the same endpoint for now - PUT `/edit-profile`

NOTE: Separating state is optional, whichever works best just proceed with it but you do need to implement the separate button and refactor the boiler plate.

File: client/src/components/profile/editprofile.js [hunk 5]
```diff
@@ -143,22 +218,34 @@ class EditProfilePopup extends Component {
...
 
   componentDidMount() {
-    const {fullname, email, gender, country} = this.props;
+    const {fullname, email, gender, country, avatar } = this.props;
+    console.log(avatar.top);
     this.setState({
       fullname,
       email,
+      avatar,
+      top: avatar.top,
```
Comment (fave77) [inline client/src/components/profile/editprofile.js]: Have these states related to avatar completely cohesive into one object called `avatar` or make the state completely isolated.

File: client/src/components/profile/editprofile.js [hunk 6]
```diff
@@ -69,6 +80,57 @@ class EditProfilePopup extends Component {
       country: e.target.value
     })
   }
+
```
Comment (fave77) [inline client/src/components/profile/editprofile.js]: These boilerplate methods should be refactored into a separate `editavatar.js` file.

File: client/src/components/profile/editprofile.js [hunk 7]
```diff
@@ -33,11 +33,22 @@ class EditProfilePopup extends Component {
 
     this.state = {
       isEditedProfileSaving: false,
-      avatar: '',
       fullname: '',
       email: '',
       gender: 'NA',
       country: 'NA',
+      avatar:'',
```
Comment (fave77) [inline client/src/components/profile/editprofile.js]: Have these states related to avatar completely cohesive into one object called avatar or make the state completely isolated.

**Review by fave77 — CHANGES_REQUESTED** [review]
@tend2infinity Good job implementing the feature as requested. However, two major changes need to be addressed before merging this PR.

- Make a separate button for `Edit Avatar`
- Don't make a separate model for `Avatar`

Although the functional requirements have been met, the code is too tightly coupled and less cohesive. For example, you are setting the default values of the prefilled avatar schema inside `user-controller.js`. In this case, you could have delegated that task to the client or create a separate service on the server.

The point is that there are multiple scenarios for structuring the code better and I have listed some of those that I could. 

Please make the necessary changes and don't change anything from a functional standpoint as this looks good except for the button.

Looking forward to your updated PR 👍🏻

**Comment by tend2infinity** [comment]
@fave77 I made a separate model because in that issue it was mentioned that I need to change the avatar type into object Id and use a ref to link another model which will contain the properties. I did that just to reduce the amount of information in the profile model but nevermind now I'll condense that in the profile model itself

File: client/src/components/profile/editprofile.js [hunk 8]
```diff
@@ -143,22 +218,34 @@ class EditProfilePopup extends Component {
   }
 
   componentDidMount() {
-    const {fullname, email, gender, country} = this.props;
+    const {fullname, email, gender, country, avatar } = this.props;
+    console.log(avatar.top);
```
Comment (apurv-1) [inline client/src/components/profile/editprofile.js]: Please don't leave these console logs on production!

File: server/src/controllers/profile.controller.js [hunk 9]
```diff
@@ -1,18 +1,22 @@
...
       return res
         .status(401)
         .json({
           success: false,
           msg: 'Could not find profile!'
         });
+        
 
     const { fullname, email, avatar, gender, country, joined } = profile;
+    console.log(profile)
```
Comment (apurv-1) [inline server/src/controllers/profile.controller.js]: and just make sure to remove these before the final merge.

**Comment by tend2infinity** [comment]
Hey, @fave77 I've made the required changes. I resolved almost all of your issues except one in which you told me to initialize the default properties of an avatar from the frontend! For this issue, I would recommend doing it from the backend itself because if we implement that from the frontend then, whenever a new user gets registered **the profile model will not contain the initial properties of the avatar in the avatar field**. That would create a problem because every time when we get to the profile page, all the information of that particular user is fetched from the **"profiles"** model and not from the frontend. We may tackle this issue by initializing an "empty string" at the time of registration of a new user and then making a check in the frontend after getting the information from the '/fetch' route that looks something like

` if (avatar.top===' ' && avatar.accessories === ' ' && avatar.hairColor === ' '&&& .....) 
{
top:"LongHairStraight",
      accessories:"Blank",
      hairColor:"BrownDark",
      facialHair:"Blank",
      clothes:"BlazerShirt",
      eyes:"Default",
      eyebrow:"Default",
      mouth:"Default",
      skin:"Light",
      clothColor: "Black"
}
`
But I feel like this kind of logic isn't the best suited, tell me what you think about it

File: server/src/controllers/user.controller.js [hunk 10]
```diff
@@ -35,7 +35,18 @@ const registerUser = async (fullname, username, password, email, status, lichess
...
+      accessories:"Blank",
+      hairColor:"BrownDark",
+      facialHair:"Blank",
+      clothes:"BlazerShirt",
+      eyes:"Default",
+      eyebrow:"Default",
+      mouth:"Default",
+      skin:"Light",
+      clothColor: "Black"
+    },
```
Comment (fave77) [inline server/src/controllers/user.controller.js:49]: This will do!

**Review by fave77 — APPROVED** [review]
@tend2infinity Looks fine to me! However, I can see you have added some additional styling so just provide a demo video for the latest changes. Will merge then 👍🏻 

Great job btw!!

**Comment by tend2infinity** [comment]
Video:

Uploading bandicam 2021-05-27 19-13-03-618.mp4…

**Comment by tend2infinity** [comment]
Video:

https://user-images.githubusercontent.com/61948033/119837411-2e34c100-bf20-11eb-94de-846f3bfd79e2.mp4

**Review by fave77 — APPROVED** [review]

**Pull request merged**